src/topology/path_connected.lean

```diff
@@ -0,0 +1,589 @@
+  { exfalso, linarith }
+end
+
+lemma range_proj_I : range proj_I = univ :=
+begin
+  rw eq_univ_iff_forall,
+  rintro ⟨t, t_in⟩,
+  use [t, proj_I_I t_in],
+end
+
+@[continuity]
+lemma continuous_proj_I : continuous proj_I :=
+begin
+  refine continuous_induced_rng' (coe : I → ℝ) rfl _,
+  have : continuous (λ t : ℝ, if t ≤ 0 then 0 else if t ≤ 1 then t else 1),
+  { refine continuous_if _ continuous_const (continuous_if _ continuous_id continuous_const) ;
+    simp [Iic_def, zero_le_one] },
+  convert this,
+  ext,
+  dsimp [proj_I],
+  split_ifs ; refl
+end
+
+variables {β : Type*}
+
+/-- Extension of a function defined on the unit interval to `ℝ`, by precomposing with
+the projection. -/
+def I_extend {β : Type*} (f : I → β) : ℝ → β :=
+f ∘ proj_I
+
+@[continuity]
+lemma continuous.I_extend {f : I → X} (hf : continuous f) : continuous (I_extend f) :=
+hf.comp continuous_proj_I
+
+lemma I_extend_extends (f : I → β) {t : ℝ} (ht : t ∈ I) : I_extend f t = f ⟨t, ht⟩ :=
+by simp [I_extend, proj_I_I, ht]
+
+@[simp] lemma I_extend_zero (f : I → β) : I_extend f 0 = f 0 :=
+I_extend_extends _ _
+
+@[simp] lemma I_extend_one (f : I → β) : I_extend f 1 = f 1 :=
+I_extend_extends _ _
+
+@[simp] lemma I_extend_range (f : I → β) : range (I_extend f) = range f :=
+begin
+  rw [I_extend, range_comp],
+  convert image_univ,
+  exact range_proj_I
+end
+
+instance : connected_space I := subtype.connected_space ⟨⟨0, by split ; norm_num⟩, is_preconnected_Icc⟩
+
+/-- Continuous path connecting two points `x` and `y` in a topological space -/
+structure path (x y : X) :=
+(to_fun : I → X)
+(continuous' : continuous to_fun)
+(left' : to_fun 0 = x)
+(right' : to_fun 1 = y)
+
+instance : has_coe_to_fun (path x y) := ⟨_, path.to_fun⟩
+
+namespace path
+
+variable (γ : path x y)
+
+@[continuity]
+protected lemma continuous : continuous γ :=
+γ.continuous'
+
+@[simp] protected lemma left : γ 0 = x :=
+γ.left'
+
+@[simp] protected lemma right : γ 1 = y :=
+γ.right'
+
+/-- The constant path from a point to itself -/
+def refl (x : X) : path x x :=
+{ to_fun := λ t, x,
+  continuous' := continuous_const,
+  left' := rfl,
+  right' := rfl }
+
+/-- The reverse of a path from `x` to `y`, as a path from `y` to `x` -/
+def symm (γ : path x y) : path y x :=
+{ to_fun      := γ ∘ σ,
+  continuous' := by continuity,
+  left'       := by simpa [-path.right] using γ.right,
+  right'      := by simpa [-path.left] using γ.left }
+
+/-- A continuous map extending a path to `ℝ`, constant before `0` and after `1`. -/
+def extend : ℝ → X := I_extend γ
+
+lemma continuous_extend : continuous γ.extend :=
+γ.continuous.I_extend
+
+@[simp] lemma extend_zero : γ.extend 0 = x :=
+by simp [extend]
+
+@[simp] lemma extend_one : γ.extend 1 = y :=
+by simp [extend]
+
+local attribute [simp] Iic_def
+
+/-- Concatenation of two paths from `x` to `y` and from `y` to `z`, putting the first
+path on `[0, 1/2]` and the second one on `[1/2, 1]`. -/
+def trans (γ : path x y) (γ' : path y z) : path x z :=
```

Comment:
Not sure if it's useful without an infix notation (is it used outside of `calc`?). Though if someone wants to use it in `calc`, a local infix notation will do.

```suggestion
@[trans] def trans (γ : path x y) (γ' : path y z) : path x z :=
```